pyteal/compiler/sourcemap.py

```diff
@@ -0,0 +1,1204 @@
+        """
+        return all(
+            [
+                not self.include_pcs or self._cached_pc_sourcemap,
+                self._cached_r3sourcemap,
+                self._cached_tmis,
+            ]
+        )
+
+    @classmethod
+    def _unexpected_error(cls, msg: str) -> TealInternalError:
+        return TealInternalError(
+            f"""{msg}
+        {cls.UNEXPECTED_ERROR_SUFFIX}"""
+        )
+
+    def build(self) -> None:
+        if self._built():
+            return
+
+        if self.include_pcs:
+            self._build_pc_sourcemap()
+
+        # Validation
+        # sanity check teal_chunks, but discard them in the rest of the computation
+        # assert teal_chunks <--1-to-1--> components
+
+        if (n := len(self.teal_chunks)) != len(self.components):
+            raise self._unexpected_error(
+                f"expected same number of teal chunks ({n}) and components ({len(self.components)})"
+            )
+
+        if n == 0:
+            raise self._unexpected_error(
+                "cannot generate empty source map: no components"
+            )
+
+        # PASS I. Deduce the Best Frame Candidate (BFC) from each individual `NatalStackFrame`
+        # See NatalStackFrame.__init__() for steps 1-7 which happen when an Expr is created
+        # 8. the list comprehension below converts each element
+        #   FROM: the component's NatalStackFrame._frames : list[StackFrame]
+        #   TO:   a PyTealFrame
+        # overall resulting in a list[PyTealFrame]
+        self._best_frames = [
+            tc.stack_frames().best().as_pyteal_frame() for tc in self.components
+        ]
+
+        if not self._best_frames:
+            raise self._unexpected_error(
+                f"This shouldn't have happened as we already checked! Check again: {len(self.components)=}"
+            )
+
+        # PASS II. Attempt to fill any "gaps" by inferring from adjacent BFC's
+        self._best_frames, inferred = self._infer(self._best_frames)
+        if inferred:
+            self._inferred_frames_at = inferred
+
+        lineno = 1
+        for i, best_frame in enumerate(self._best_frames):
+            teal_chunk = self.teal_chunks[i]
+            for line in teal_chunk.splitlines():
+                pcsm = cast(PCSourceMap, self._cached_pc_sourcemap)
+                pcs = None
+                if self.include_pcs:
+                    pcs = pcsm.line_to_pc.get(lineno - 1, [])
+                self._cached_tmis.append(
+                    TealMapItem(
+                        pt_frame=best_frame,
+                        teal_lineno=lineno,
+                        teal_line=line,  # type: ignore
+                        teal_component=self.components[i],
+                        pcs=pcs,
+                    )
+                )
+                lineno += 1
+
+        self._build_r3sourcemap()
+
+        if not NatalStackFrame._debug:
+            self._best_frames = []
+            self._inferred_frames_at = []
```

Comment:
nit: you can try 
```suggestion
            self._best_frames.clear()
            self._inferred_frames_at.clear()
```
to release the memory immediately, see https://blog.finxter.com/list-clear-vs-new-list/

Reply:
Great catch. I was inspired to try no. 4 on [this S.O.](https://stackoverflow.com/questions/850795/different-ways-of-clearing-lists#answer-44349418)

b38a83c29d61e1a218dfda82247de3f4d7adbde9